src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs

```diff
@@ -0,0 +1,330 @@
+using System.Runtime.InteropServices;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib decompression API.
+    /// </summary>
+    internal sealed class WebSocketInflater : IDisposable
+    {
+        internal const int FlushMarkerLength = 4;
+        internal static ReadOnlySpan<byte> FlushMarker => new byte[] { 0x00, 0x00, 0xFF, 0xFF };
+
+        private ZLibStreamHandle? _stream;
+        private readonly int _windowBits;
+        private readonly bool _persisted;
+
+        /// <summary>
+        /// There is no way of knowing, when decoding data, if the underlying deflater
+        /// has flushed all outstanding data to consumer other than to provide a buffer
+        /// and see whether any bytes are written. There are cases when the consumers
+        /// provide a buffer exactly the size of the uncompressed data and in this case
+        /// to avoid requiring another read we will use this field.
+        /// </summary>
+        private byte? _remainingByte;
+
+        /// <summary>
+        /// When the inflater is persisted we need to manually append the flush marker
+        /// before finishing the decoding.
+        /// </summary>
+        private bool _needsFlushMarker;
+
+        private byte[]? _buffer;
+
+        /// <summary>
+        /// The position for the next unconsumed byte in the inflate buffer.
+        /// </summary>
+        private int _position;
+
+        /// <summary>
+        /// How many unconsumed bytes are left in the inflate buffer.
+        /// </summary>
+        private int _available;
+
+        internal WebSocketInflater(int windowBits, bool persisted)
+        {
+            Debug.Assert(windowBits >= 9 && windowBits <= 15);
+
+            // We use negative window bits to instruct deflater to expect raw deflate data
+            _windowBits = -windowBits;
+            _persisted = persisted;
+        }
+
+        /// <summary>
+        /// Indicates that there is nothing left for inflating. If there is
+        /// more data left to be received for the message then call <see cref="Initialize(long, int)"/>
+        /// and receive directly into <see cref="Memory"/>.
+        /// </summary>
+        public bool Finished { get; private set; } = true;
+
+        public Memory<byte> Memory => _buffer;
+
+        public Span<byte> Span => _buffer;
+
+        public void Dispose()
+        {
+            _stream?.Dispose();
+            ReleaseBuffer();
+        }
+
+        /// <summary>
+        /// Initializes the inflater by allocating a buffer so the websocket can receive directly onto it.
+        /// </summary>
+        /// <param name="payloadLength">the length of the message payload</param>
+        /// <param name="clientBufferLength">the length of the buffer where the payload will be inflated</param>
+        public void Initialize(long payloadLength, int clientBufferLength)
```

Comment:
nit: rename clientBuffer to userBuffer. I think you in general want to avoid using what the client told you (the payload length)